src/injector/inject.ts

```diff
@@ -1,57 +1,79 @@
 }
 
 interface ChibiCompatibleVM extends VM {
     ccExtensionManager?: {
-        info: Record<string, {
-            api: number;
-        }>;
-        getExtensionLoadOrder (extensions: string[]): unknown;
+        info: Record<
+            string,
+            {
+                api: number;
+            }
+        >;
+        getExtensionLoadOrder(extensions: string[]): unknown;
     };
     setLocale?: (locale: string, ...args: unknown[]) => unknown;
+    getLocale?: () => string;
 }
 
 const MAX_LISTENING_MS = 30 * 1000;
 
-
-function getBlocklyInstance () {
-    const elem = document.querySelector('[class^="gui_blocks-wrapper"]');
-    if (!elem) return null;
-    const internalKey = Object.keys(elem).find(
-        (key) => key.startsWith('__reactInternalInstance$') ||
-        key.startsWith('__reactFiber$')
-    );
-    if (!internalKey) return;
-    // @ts-expect-error
-    const internal = elem[internalKey];
-    let childable = internal;
-    try {
-        while (((childable = childable.child), !childable || !childable.stateNode || !childable.stateNode.ScratchBlocks)) {}
-    } catch (e: unknown) {
-        return;
+/**
+ * Get Blockly instance.
+ * @param vm Virtual machine instance. For some reasons we cannot use VM here.
+ * @returns Blockly instance.
+ */
+function getBlocklyInstance(vm: any): any | undefined {
+    // hijack Function.prototype.apply to get React element instance.
+    function hijack(fn: (...args: unknown[]) => void): any {
+        const _orig = Function.prototype.apply;
+        Function.prototype.apply = function (thisArg: any) {
+            return thisArg;
+        };
+        const result = fn();
+        Function.prototype.apply = _orig;
+        return result;
     }
-    return childable?.stateNode.ScratchBlocks;
+    const events = vm._events?.EXTENSION_ADDED;
+    if (events) {
+        if (events instanceof Function) {
+            // it is a function, just hijack it.
+            const result = hijack(events);
+            if (result && result.ScratchBlocks) return result.ScratchBlocks;
+        } else {
+            // it is an array, hijack every listeners.
+            for (const value of events) {
+                const result = hijack(value);
+                if (result && result.ScratchBlocks) return result.ScratchBlocks;
+            }
+        }
+    }
+    return undefined; // Method failed.
 }
-
-export function trap () {
+/**
+ * Trap to get Virtual Machine instance.
+ * @param open window.open function (compatible with ccw).
+ * @return Callback promise. After that you could use window.chibi.vm to get the virtual machine.
+ */
+export function trap(open: typeof window.open): Promise<void> {
     window.chibi = {
         // @ts-expect-error defined in webpack define plugin
         version: __CHIBI_VERSION__,
         registeredExtension: {},
-        openFrontend: openFrontend
+        openFrontend: openFrontend.bind(null, open),
```

Comment:
😎

Reply:
Simple curry function to ensure compatibility.